apluslms_roman/backends/__init__.py

```diff
@@ -26,21 +27,32 @@ class BuildStep:
     cmd: If not None, docker command
     mnt: If not None, course data is mounted to this path in RW mode
     env: If not None, dict that is given as environment for the image
+    ref: Name/index of the step
     """
-    __slots__ = ('img', 'cmd', 'mnt', 'env')
+    __slots__ = ('img', 'cmd', 'mnt', 'env', 'name', 'ref')
 
     @classmethod
-    def from_config(cls, data):
-        if isinstance(data, dict):
+    def from_config(cls, index, data):
+        if isinstance(data, Mapping):
             if 'img' not in data:
                 raise RuntimeError("Missing image name (img) in step configuration: {}".format(data))
             img = clean_image_name(data['img'])
-            return cls(img, data.get('cmd'), data.get('mnt'), data.get('env'))
+            return cls(
+                img,
+                data.get('cmd'),
+                data.get('mnt'),
+                dict(data.get('env', {})),
+                data.get('name'),
+                index
+            )
         else:
-            return cls(clean_image_name(data), None, None, None)
+            return cls(clean_image_name(data), None, None, None, None)
 
-    def __init__(self, img, cmd, mnt, env):
-        self.img, self.cmd, self.mnt, self.env = img, cmd, mnt, env
+    def __init__(self, img, cmd, mnt, env, name, ref):
```

Comment:
We could reorder these to be `ref, img, cmd=None, mnt=None, env=None,, name=None`.
And when setting `self.name` we could default that to `ref` if no name is provided. (`self.name = name or ref`)

Reply:
currently when listing steps I don't list a step name if step is none. if name is ref then that doesn't work and the only way to check if the name is ref is to compare it with the ref (or check if name is int).

Reply:
Good point. Guess it's better have name as None when not set. `__str__` can then use name or ref if there is no name.